crypto/crypto/src/ed25519.rs

```diff
@@ -76,6 +77,24 @@ impl Clone for Ed25519PrivateKey {
 #[derive(DeserializeKey, Clone, SerializeKey)]
 pub struct Ed25519PublicKey(ed25519_dalek::PublicKey);
 
+/// A private tag type that tracks if a public key is checked to protect against invalid point
+/// attacks, small subgroup attacks, and typos. This tag type is only used at compilation time.
+/// This type should remain private.
+#[cfg(feature = "const_generics")]
+type ValidPublicKeyTag = ValidPublicKeyTagKind<VALID_PUBLIC_KEY_TAG_MASK>;
+
+#[cfg(not(feature = "const_generics"))]
+struct ValidPublicKeyTag {}
+
+/// A generic tag type intended to only be used by ValidPublicKeyTag
+#[cfg(feature = "const_generics")]
+struct ValidPublicKeyTagKind<const MASK: TagPropagationSet> {}
+
+/// The propagation set of ValidPublicKeyTag. An empty propagation is used to make sure that
+/// ValidPublicKeyTag can only be explicitly attached to public keys.
+#[cfg(feature = "const_generics")]
+const VALID_PUBLIC_KEY_TAG_MASK: TagPropagationSet = tag_propagation_set!();
```

Comment:
not sure I understand what this really does :D

Reply:
It prevents propagating the tag (which presence marks a checked public key) through e.g. equality. 
Though in all rigor, cloning and equality should propagate `ValidPublicKeyTag`.
 
https://github.com/facebookexperimental/MIRAI/blob/master/documentation/TagAnalysis.md#declaring-tags

Reply:
Ah I see, makes sense! From a quick glance I don't see a cloning here though: https://github.com/facebookexperimental/MIRAI/blob/8f0596f02b557c3e357767ee73c437d4a9bf8c5c/annotations/src/lib.rs#L33

Reply:
It's a bit unfortunate that cloning is not yet handled properly (see this [issue](https://github.com/facebookexperimental/MIRAI/issues/611)), because essentially tags are stored as a model field of a struct, which is not transferred by cloning. I'm investigating this problem.

Reply:
equal means copy?

Reply:
Equal means equality checks. From the perspective of semantics, copying generates duplicates of a value, so it is assumed to always propagate tags.

Reply:
Cloning now tracks tag/model fields properly (see this [PR](https://github.com/facebookexperimental/MIRAI/pull/613)).